url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/sts/impl/NusmvSTS.java

```diff
@@ -0,0 +1,131 @@
+package eu.fbk.soa.evolution.sts.impl;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import eu.fbk.soa.evolution.sts.Action;
+import eu.fbk.soa.evolution.sts.STS;
+import eu.fbk.soa.evolution.sts.State;
+import eu.fbk.soa.evolution.sts.Transition;
+
+
+
+
+public class NusmvSTS extends AbstractSTS implements STS {
+	
+	private Set<Action> actions;
+	
+	public NusmvSTS() {
+		this.states = new HashSet<State>();
+		this.actions = new HashSet<Action>();
+		this.transitions = new HashSet<Transition>();
+	}
+
+	public NusmvState addState(Map<String, String> varValues) {		
+		NusmvState newState = new NusmvState(varValues);
+		for (State state : states) {
+			if (state.equals(newState))
+				return (NusmvState) state;
+		}
+		this.states.add(newState);
+		return newState;
+	}
+
+	public Action addAction(Map<String, String> varValues) {
+		Action newAction = new NusmvAction(varValues);
+		for (Action act : actions) {
+			if (act.equals(newAction))
+				return act;
+		}
+		this.actions.add(newAction);
+		return newAction;
+	}
+	
+	public State getInitialState() {
+		for (State state : states) {
+			boolean isInitial = true;
+			for (Transition t : transitions) {
+				if (t.getTarget().equals(state)) {
+					isInitial = false;
+					break;
+				}
+			}
+			if (isInitial) {
+				return state;
+			}
+		}
+		return null;
+	}
+
+	public Set<Action> getActions() {
+		return this.actions;
+	}
+	
+	public NusmvSTS getCopy() {
+		NusmvSTS cloneSTS = new NusmvSTS();
+		cloneSTS.addStates(this.states);
+		cloneSTS.addTransitions(this.transitions);
+		cloneSTS.addActions(this.actions);
+		return cloneSTS;
+	}
+
+	public void addActions(Collection<Action> actionCollection) {
+		this.actions.addAll(actionCollection);		
+	}
+
+	@Override
+	public void addAction(Action newAction) {
+		for (Action act : actions) {
+			if (act.equals(newAction)) {
+				return;
+			}
+		}
+		this.actions.add(newAction);
+	}
+
+	@Override
+	public Set<Action> getInputActions() {
+		Set<Action> inputActions = new HashSet<Action>();
+		for (Action act : this.actions) {
+			if (act.isInputAction()) {
+				inputActions.add(act);
+			}
+		}
+		return inputActions;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 90:96, 101:107.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.